### PR TITLE
Rust & Typescript Cursors for the CST

### PR DESCRIPTION
**Pull request by AntonyBlakey**
Does not include TS Visitor - waiting for Omar's TS/Rust codegen unification PR to land before completing this task.

### CODE REVIEW

File: .changeset/cuddly-avocados-sing.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"changelog": minor
+---
+
+Add a Rust Cursor API and refactor the Rust Visitor API to run on top of it. Add first pass of Typescript binding to the Cursor API, but no TS Visitor yet. Add TokenNode.text to the TS API.
```
Comment (OmarTawfik) [inline .changeset/cuddly-avocados-sing.md:5]: I wonder if this should split into three entires (3 changesets)?

- Add a Rust Cursor API and refactor the Rust Visitor API to run on top of it.
- Add first pass of Typescript binding to the Cursor API, but no TS Visitor yet.
- Add `TokenNode.text` to the TS API.

File: crates/codegen/syntax_templates/src/shared/cst.rs [hunk 2]
```diff
@@ -52,4 +53,34 @@ impl Node {
             Self::Token(node) => (&node.text).into(),
         }
     }
+
+    pub fn cursor(&self) -> Cursor {
+        Cursor::new(self.clone())
+    }
+
+    pub fn is_token_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cst.rs:61]: For `is_token_kind` and `is_rule_kind`, I wonder why do they take multiple kinds, rather than only one? is it an optimization for NAPI? does that mean that we won't do that if it was a native TS API?

File: crates/codegen/syntax_templates/src/shared/cst.rs [hunk 3]
```diff
@@ -52,4 +53,34 @@ impl Node {
...
+        if let Node::Token(token_node) = self {
+            if kinds.contains(&token_node.kind) {
+                return Some(token_node);
+            }
+        }
+        return None;
+    }
+
+    pub fn is_token_matching<F: Fn(&Rc<TokenNode>) -> bool>(&self, predicate: F) -> bool {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cst.rs:70]: Is `is_token_matching` and `is_rule_matching` intended to be shared between languages? what is the use case here?

File: crates/codegen/syntax_templates/src/shared/cursor.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorLeaf {
+    node: Node,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+impl CursorLeaf {
+    pub fn text_range(&self) -> TextRange {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cursor.rs:24]: I wonder why are we using `text_range` here vs just `range` on nodes?

File: crates/codegen/syntax_templates/src/shared/cursor.rs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+                child_number: 0,
+                text_offset: Default::default(),
+            }),
+        }
+    }
+
+    // Unlike clone, this re-roots at the current node.
+    // It does preserve the correct text offset however,
+    // even though the path is reset.
+    pub fn spawn(&self) -> Self {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cursor.rs:52]: thinking out loud, is `spawn` common terminology for cursors? vs, for example, `fork`?

File: crates/codegen/syntax_templates/src/shared/cursor.rs [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+            path: vec![],
+            leaf: self.leaf.clone(),
+        }
+    }
+
+    pub fn is_finished(&self) -> bool {
+        self.leaf.is_none()
+    }
+
+    pub fn node_or_none(&self) -> Option<Node> {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cursor.rs:63]: regarding `node()` and `node_or_none()`, I might be missing something, but I would have expected the rust idiomatic way is to return `Option<Node>` only/directly for things like this, and let the user `expect()`/`unwrap()` if needed? why the extra API?

File: crates/codegen/syntax_templates/src/shared/cursor.rs [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    pub fn text_range(&self) -> TextRange {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it hasn no text_range")
+            .text_range()
+    }
+
+    pub fn path_rule_nodes(&self) -> Vec<Rc<RuleNode>> {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cursor.rs:89]: not sure I understand the use case for `path_rule_nodes()`. I see it is currently unused?

File: crates/codegen/syntax_templates/src/shared/cursor.rs [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+                return false;
+            }
+        }
+        while self.go_to_last_child() {}
+        return true;
+    }
+
+    pub fn go_to_parent(&mut self) -> bool {
+        self.leaf = self.path.pop().map(|path_element| CursorLeaf {
+            node: Node::Rule(path_element.rule_node),
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cursor.rs:124]: IIUC, that means that when there is an empty path with an existing leaf, this method will delete the leaf and return false. If there is no parent (path is empty), shouldn't we return false without modifying the current cursor?

File: crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        Self::new(self.0.spawn())
+    }
+
+    #[napi(getter)]
+    pub fn is_finished(&self) -> bool {
+        self.0.is_finished()
+    }
+
+    #[napi(getter, ts_return_type = "RuleNode | TokenNode | null")]
+    pub fn node(&self, env: Env) -> Option<JsObject> {
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs:30]: I think I'm missing something. This would convert it to a plain JS object, however, us typing it with `RuleNode | TokenNode` means that it will have the usual APIs on it, like the ones that need round tripping to Rust. How does that work?

this is related to other `to_js()` uses as well.

File: crates/codegen/syntax_templates/src/shared/visitor.rs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        node: &Rc<RuleNode>,
+        cursor: &Cursor,
+    ) -> Result<VisitorEntryResponse, E> {
+        Ok(VisitorEntryResponse::StepIn)
+    }
+
+    fn rule_exit(
+        &mut self,
+        node: &Rc<RuleNode>,
+        cursor: &Cursor,
```
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/visitor.rs:18]: TOL, I wonder if we should be exposing the cursor to the visitor API, even if it cannot be used/mutated by the caller? what is the use case here?

File: crates/solidity/outputs/npm/package/syntax/nodes/index.js [hunk 11]
```diff
@@ -6,6 +6,7 @@
 
 const generated = require("../../generated");
 
+module.exports.Cursor = generated.Cursor;
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/package/syntax/nodes/index.js:9]: I suggest moving `Cursor` to `syntax/visitors` namespace to match the other export.

File: crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs [hunk 12]
```diff
@@ -24,21 +24,19 @@ struct TestNodeBuilder {
 }
 
 impl Visitor<()> for TestNodeBuilder {
-    fn enter_rule(
+    fn rule_enter(
```
Comment (OmarTawfik) [inline crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs:27]: I wonder why the change from `enter_rule()` to `rule_enter()`?

File: crates/solidity/outputs/cargo/tests/src/doc_examples/cursor_api.rs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+use slang_solidity::{
+    language::Language,
+    syntax::{
+        nodes::{RuleKind, TokenKind},
+        parser::ProductionKind,
+    },
+};
+
+#[test]
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/tests/src/doc_examples/cursor_api.rs:12]: I suggest splitting into three tests for readability.

File: .changeset/cuddly-avocados-sing.md (hunk 1 above)
Comment (AntonyBlakey) [reply .changeset/cuddly-avocados-sing.md:5]: done

File: crates/codegen/syntax_templates/src/shared/cst.rs (hunk 2 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cst.rs:61]: That is for uniformity with the Cursor API, where you can search for more than one at a time, without having to do multiple searches over multiple Cursors. In TS it will be `is_token_kind(TokenKind, ...)`. 

Note that it also returns the matched TokenNode, meaning no `match` is required, and it functions therefore as a typeguard.

Note also the following `is_token_matching` which is also for alignment with the cursor API.

File: crates/codegen/syntax_templates/src/shared/cst.rs (hunk 3 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cst.rs:70]: Yes.

The use case if to supply a predicate, which is passed the TokenNode or RuleNode i.e. pre-matched. For use when traversing, and for alignment with the Cursor API

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 8 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:124]: I went back and forth about this. It is assumed in methods that call `go_to_parent()` that failure of that method also completes the cursor. I didn't like that one navigator having different behaviour. Since you raised it, I am swayed and I will add an explicit `complete()` method in the callers.

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 7 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:89]: So that you can access the parent `RuleNode`s when using a cursor to navigate the tree.

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 6 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:63]: Because there are many cases where the client know that node isn't none e.g. after a successful `go_to_*`, and I wanted to not require `.unwrap()` there, and instead catch it in one place. Happy to change the method names if you have a suggestion?

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 5 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:52]: It is for me. `fork()` is overloaded I think.

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 4 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:24]: I prefer the standardized name, so I will make that change.

In this case it's only internal because CursorLeaf isn't part of the public API.

File: crates/codegen/syntax_templates/src/shared/visitor.rs (hunk 10 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/visitor.rs:18]: For two reasons
1. You can start a `Visitor` at any `Cursor` position. It is nice in the `Visitor` to be able to get to the `Cursor` and do some more internal iteration. Apart from being convenient, it makes internal and external iterators nicely symmetric.
2. Regardless of the symmetry, it is convenient in the visitor to be able to access things like the `text_range` from the `Cursor`, and also the `path_rule_nodes`. As the `Cursor` API is extended, the `Visitor` implementations can use those facilities.

File: crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs (hunk 9 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs:30]: In this case we have to return `JsObject` because the result is polymorphic. And fundamentally, `JsObject` is what is passed even if the Rust type were explicit.

File: crates/solidity/outputs/cargo/tests/src/doc_examples/cursor_api.rs (hunk 13 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/cargo/tests/src/doc_examples/cursor_api.rs:12]: I thought about that, but because the 'test' nature of this is not actually for testing, but just to validate that the example is ok, I decided that the clarity I could gain by simply repeating a section of code with exactly the same result but using different API patterns was worth having the repeated functionality.

File: crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs (hunk 12 above)
Comment (AntonyBlakey) [reply crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs:27]: Because I removed the `enter||exit_` prefix from `token`, and I wanted the method names in the `Visitor` to start with the type name i.e. `rule|token` for regularity.

File: crates/solidity/outputs/npm/package/syntax/nodes/index.js (hunk 11 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/npm/package/syntax/nodes/index.js:9]: I would rather have cursor and visitor in nodes, so I think I'll change the other export

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 8 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:124]: Actually, now that I think about it, it needs to be like this. The `go_to_*` methods fall into two categories:
1. Those that fail without moving, which are only the `go_to_*_sibling` methods; and
2. Those that fail and complete the cursor in the process. Which is all others - and those all fail via `go_to_parent`, which is the only place that completes the cursor.

We have to complete the cursor so that the start and end states are distinguishable, otherwise the cursor becomes cyclic. I've built cursor APIs like that before, and always regretted it. The DX is awful.

File: crates/solidity/outputs/cargo/tests/src/doc_examples/cursor_api.rs (hunk 13 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/cargo/tests/src/doc_examples/cursor_api.rs:12]: I think it is also for "testing", since the Rust test infra is the same in all projects, this is a quick way for someone to just copy a test locally, and edit/run it for experimenting.

It also indicates to readers that these are separate, instead of having side effects on later statements.

File: crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs [hunk 14]
```diff
@@ -17,40 +16,38 @@ struct ContractCollector {
...
         node: &Rc<RuleNode>,
-        _path: &Vec<Rc<RuleNode>>,
-        _range: &TextRange,
+        _cursor: &Cursor,
     ) -> Result<VisitorEntryResponse> {
-        if node.kind != RuleKind::ContractDefinition {
-            return Ok(VisitorEntryResponse::StepIn);
+        if node.kind == RuleKind::ContractDefinition {
+            if let Node::Token(token) = &node.children[2] {
+                assert_eq!(token.kind, TokenKind::Identifier);
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs:26]: nit: `ensure!()` to be consistent with `bail!()` instead of panicking?

File: crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs [hunk 15]
```diff
@@ -17,40 +16,38 @@ struct ContractCollector {
...
+    fn rule_enter(
         &mut self,
         node: &Rc<RuleNode>,
-        _path: &Vec<Rc<RuleNode>>,
-        _range: &TextRange,
+        _cursor: &Cursor,
     ) -> Result<VisitorEntryResponse> {
-        if node.kind != RuleKind::ContractDefinition {
-            return Ok(VisitorEntryResponse::StepIn);
+        if node.kind == RuleKind::ContractDefinition {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs:24]: nit: I suggest quitting early when possible (like the existing visitor) for readability, and to decrease nesting. But not a blocker at all.

```rust
if !interesting {
  return;
}
do_something();
```

rather than:

```rust
if interesting {
  do_something();
}
```

File: crates/solidity/outputs/npm/package/generated/index.d.ts [hunk 16]
```diff
@@ -623,12 +623,35 @@ export class RuleNode {
...
+  goToNext(): boolean;
+  goToNextNonDescendent(): boolean;
+  goToPrevious(): boolean;
+  goToParent(): boolean;
+  goToFirstChild(): boolean;
+  goToLastChild(): boolean;
+  goToNthChild(childNumber: number): boolean;
+  goToNextSibling(): boolean;
+  goToPreviousSibling(): boolean;
+  findRuleKind(kinds: Array<RuleKind>): RuleNode | null;
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/package/generated/index.d.ts:651]: TOL: thoughts about unifying `findXXXX()` and `findXXXXWithoutRecursing()` to a default argument `findXXXX(recursive = true)`?

File: crates/codegen/syntax_templates/src/shared/cst.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/cst.rs:61]: I think `is_XX()` usually indicates returning booleans, a cheap check that can be chained in a boolean condition with other conditions. Thoughts about something like `as_token()`? which indicates a "cast" that can succeed/fail (thus returning an option).

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 6 above)
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:63]: I think this is idiomatic to let the user make that decision, and to make it explicit at the call site (a strong direction in rust). I think similar to how a vector/map/set would offer `get() -> Option<T>` without (the often convenient) `get_and_unwrap() -> T`

File: crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs (hunk 9 above)
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs:35]: So on the receiver side, working in NodeJS land, they would get a plain JS object.
It wouldn't have the native Rust APIs? not sure how would calling its methods would round trip.

... I think I will wait for this to be ready and play with it for a bit to understand more.

**Review by OmarTawfik — APPROVED** [review]
I've a few open questions, but will wait until the docs are ready to have a better understanding of the proposed API before continuing.

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 6 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:63]: `trait Index` provides `[]` for this, and it is implemented on `Vec` and `HashMap` for example.

File: crates/codegen/syntax_templates/src/shared/cst.rs (hunk 2 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cst.rs:61]: agreed, done

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 8 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:124]: now more obvious, with explicit completion

File: crates/codegen/syntax_templates/src/shared/cursor.rs (hunk 6 above)
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cursor.rs:63]: But in any case, solved with an explicit completed state. Also allows the cursor to be reset

File: crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs (hunk 15 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs:24]: As a functional programming fan, I prefer to have a single control flow, and therefore not use guards as often.

File: crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs (hunk 14 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/cargo/tests/src/doc_examples/visitor_api.rs:26]: done

File: crates/solidity/outputs/npm/package/generated/index.d.ts (hunk 16 above)
Comment (AntonyBlakey) [reply crates/solidity/outputs/npm/package/generated/index.d.ts:651]: I've removed the without recursing forms, but in any case I wouldn't want to add a recursive param to findXXX if the use cases were not equaly distributed, which in this case I suspect they wouldn't be.
